app/src/main/java/org/simple/clinic/setup/SetupActivityEffectHandler.kt

```diff
@@ -13,19 +12,39 @@ import org.simple.clinic.util.Optional
 import org.simple.clinic.util.scheduler.SchedulersProvider
 import org.simple.clinic.util.toOptional
 
-object SetupActivityEffectHandler {
+class SetupActivityEffectHandler(
+    private val onboardingCompletePreference: Preference<Boolean>,
+    private val uiActions: UiActions,
+    private val userDao: User.RoomDao,
+    private val appConfigRepository: AppConfigRepository,
+    private val fallbackCountry: Country,
+    private val schedulersProvider: SchedulersProvider
+) {
 
-  fun create(
-      onboardingCompletePreference: Preference<Boolean>,
-      uiActions: UiActions,
-      userDao: User.RoomDao,
-      appConfigRepository: AppConfigRepository,
-      fallbackCountry: Country,
-      schedulersProvider: SchedulersProvider
-  ): ObservableTransformer<SetupActivityEffect, SetupActivityEvent> {
+  companion object {
+    fun create(
+        onboardingCompletePreference: Preference<Boolean>,
+        uiActions: UiActions,
+        userDao: User.RoomDao,
+        appConfigRepository: AppConfigRepository,
+        fallbackCountry: Country,
+        schedulersProvider: SchedulersProvider
+    ): ObservableTransformer<SetupActivityEffect, SetupActivityEvent> {
+      return SetupActivityEffectHandler(
+          onboardingCompletePreference = onboardingCompletePreference,
+          uiActions = uiActions,
+          userDao = userDao,
+          appConfigRepository = appConfigRepository,
+          fallbackCountry = fallbackCountry,
+          schedulersProvider = schedulersProvider
+      ).build()
+    }
+  }
+
+  private fun build(): ObservableTransformer<SetupActivityEffect, SetupActivityEvent> {
     return RxMobius
         .subtypeEffectHandler<SetupActivityEffect, SetupActivityEvent>()
-        .addTransformer(FetchUserDetails::class.java, fetchUserDetails(onboardingCompletePreference, userDao, appConfigRepository, schedulersProvider.io()))
+        .addTransformer(FetchUserDetails::class.java, fetchUserDetails())
```

Comment:
Pass in the `Scheduler` probably?

Reply:
What benefit does it provide? We needed to do it earlier because the effect handler was a factory function, but now everything is a part of the class properties. This is anyway a private function so it does not need to be exposed for testing.

Reply:
Looking at the subtype effect handler tells us what scheduler the effect handler is being executed?

Reply:
Is that really important? The functions are quite small and easy to parse. Additionally, this reasoning breaks if we need to pass two schedulers to the effect handler for different use cases because we will know that the effect handler uses these two, but not for what.

My opinion on this is that if someone is at a point where they need to know what scheduler an effect handler is running on, they will be anyway looking closely at the function body and it is quite obvious which scheduler is being used there and where.

Does this make sense?

Reply:
In addition, I feel that adding the scheduler as a method parameter is exposing an implementation detail.

Reply:
Discussed offline: Decided to pass the scheduler as a parameter to the effect handler function to maintain consistency with how we'll use schedulers when adding effect handler consumers and actions.